price-feeder/oracle/provider/coinbase.go

```diff
@@ -0,0 +1,443 @@
+	}
+
+	// CoinbaseMatchResponse defines the response body for coinbase trades.
+	CoinbaseTradeResponse struct {
+		Type      string `json:"type"`       // "last_match" or "match"
+		ProductID string `json:"product_id"` // ex.: ATOM-USDT
+		Time      string `json:"time"`       // Time in format 2006-01-02T15:04:05.000000Z
+		Size      string `json:"size"`       // Size of the trade ex.: 10.41
+		Price     string `json:"price"`      // ex.: 14.02
+	}
+
+	// CoinbaseTrade defines the trade info we'd like to save.
+	CoinbaseTrade struct {
+		ProductID string // ex.: ATOM-USDT
+		Time      int64  // Time in unix epoch ex.: 164732388700
+		Size      string // Size of the trade ex.: 10.41
+		Price     string // ex.: 14.02
+	}
+
+	// CoinbaseTicker defines the ticker info we'd like to save.
+	CoinbaseTicker struct {
+		ProductID string `json:"product_id"` // ex.: ATOM-USDT
+		Price     string `json:"price"`      // ex.: 523.0
+		Volume    string `json:"volume_24h"` // 24-hour volume
+	}
+
+	// CoinbaseErrResponse defines the response body for errors.
+	CoinbaseErrResponse struct {
+		Type   string `json:"type"`   // should be "error"
+		Reason string `json:"reason"` // ex.: "tickers" is not a valid channel
+	}
+)
+
+// NewCoinbaseProvider creates a new CoinbaseProvider.
+func NewCoinbaseProvider(ctx context.Context, logger zerolog.Logger, pairs ...types.CurrencyPair) (*CoinbaseProvider, error) {
+	wsURL := url.URL{
+		Scheme: "wss",
+		Host:   coinbaseHost,
+	}
+
+	wsConn, _, err := websocket.DefaultDialer.Dial(wsURL.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to Coinbase websocket: %w", err)
+	}
+
+	provider := &CoinbaseProvider{
+		wsURL:           wsURL,
+		wsClient:        wsConn,
+		logger:          logger.With().Str("provider", "coinbase").Logger(),
+		reconnectTimer:  time.NewTicker(coinbasePingCheck),
+		trades:          map[string][]CoinbaseTrade{},
+		tickers:         map[string]CoinbaseTicker{},
+		subscribedPairs: map[string]types.CurrencyPair{},
+	}
+	provider.wsClient.SetPongHandler(provider.pongHandler)
+
+	if err := provider.SubscribeCurrencyPairs(pairs...); err != nil {
+		return nil, err
+	}
+
+	go provider.handleReceivedMessages(ctx)
+
+	return provider, nil
+}
+
+// GetTickerPrices returns the tickerPrices based on the saved map.
+func (p *CoinbaseProvider) GetTickerPrices(pairs ...types.CurrencyPair) (map[string]TickerPrice, error) {
+	tickerPrices := make(map[string]TickerPrice, len(pairs))
+
+	for _, currencyPair := range pairs {
+		price, err := p.getTickerPrice(currencyPair)
+		if err != nil {
+			return nil, err
+		}
+
+		tickerPrices[currencyPair.String()] = price
+	}
+
+	return tickerPrices, nil
+}
+
+// GetCandlePrices returns candles based off of the saved trades map.
+// Candles need to be cut up into one-minute intervals.
+func (p *CoinbaseProvider) GetCandlePrices(pairs ...types.CurrencyPair) (map[string][]CandlePrice, error) {
```

Comment:
Yeah, this looks pretty reasonable to me. The only other clean approach I can think of is to bucket all trades that are within 1 minute of each other (after sorting ofc).